### DISCUSSION TITLE
Dev Container vs. Prod Container Mismatch

### DISCUSSION THREAD

**Discussion by leocavalcante** [devcontainers/community#123] [General]
Is that a concern?

With Dev Container we have a container built for development purposes, but when it comes to a Prod Container, a Container that is meant to run the application in production environments.

Do you have two different `Dockerfile`s? How do you keep/maintain both containers using the same base/versions/extensions?

Do you use multi-stage in a single `Dockerfile`?

How do you share a single `Dockerfile` to act as the Dev Container for multiple projects, but keep a production `Dockerfile` for each project and still deal with mismatches?

Did you folks get what I'm talking about?

**Comment by samruddhikhandale** [comment]
Hi 👋 

Thanks for raising this issue and yes, I understand your concerns. You're looking for a way to maintain consistency between development and production Docker containers while also sharing a common development Dockerfile across multiple projects. Here's a general approach:

- Use multi-stage builds in a single Dockerfile: This allows you to use the same base image and extensions for both development and production stages. The development stage can include additional tools and configurations needed for development, while the production stage includes only what's necessary to run the application.

- Parameterize your Dockerfile: You can use build arguments (ARG) in your Dockerfile to make it more flexible. This way, you can pass in different values for these arguments when building for development or production.

- Use a base Dockerfile for common setup: If multiple projects share a common setup, you can create a base Dockerfile that includes these common steps. Each project's Dockerfile can then start with FROM this base image and add project-specific configurations.

- Use Docker Compose for development environments: Docker Compose allows you to define and run multi-container Docker applications, which can be useful for setting up development environments. You can include the build context and Dockerfile in the compose file, allowing you to use different Dockerfiles for different services if necessary.

Hope this helps!
